timelock-impl/src/main/java/com/palantir/atlasdb/backup/AtlasRestoreResource.java

```diff
@@ -47,31 +51,47 @@
 
     private final Function<String, AsyncTimelockService> timelockServices;
     private final ConjureResourceExceptionHandler exceptionHandler;
+    private final Supplier<BearerToken> permittedToken;
 
     @VisibleForTesting
     AtlasRestoreResource(
-            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
+            Supplier<BearerToken> permittedToken,
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        this.permittedToken = permittedToken;
         this.exceptionHandler = new ConjureResourceExceptionHandler(redirectRetryTargeter);
         this.timelockServices = timelockServices;
     }
 
     public static UndertowService undertow(
-            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
-        return AtlasRestoreClientEndpoints.of(new AtlasRestoreResource(redirectRetryTargeter, timelockServices));
+            Supplier<BearerToken> permittedToken,
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        return AtlasRestoreClientEndpoints.of(
+                new AtlasRestoreResource(permittedToken, redirectRetryTargeter, timelockServices));
     }
 
     public static AtlasRestoreClient jersey(
-            RedirectRetryTargeter redirectRetryTargeter, Function<String, AsyncTimelockService> timelockServices) {
-        return new JerseyAtlasRestoreClientAdapter(new AtlasRestoreResource(redirectRetryTargeter, timelockServices));
+            Supplier<BearerToken> permittedToken,
+            RedirectRetryTargeter redirectRetryTargeter,
+            Function<String, AsyncTimelockService> timelockServices) {
+        return new JerseyAtlasRestoreClientAdapter(
+                new AtlasRestoreResource(permittedToken, redirectRetryTargeter, timelockServices));
     }
 
     @Override
     public ListenableFuture<CompleteRestoreResponse> completeRestore(
             AuthHeader authHeader, CompleteRestoreRequest request) {
-        return handleExceptions(() -> completeRestoreInternal(request));
+        return handleExceptions(() -> completeRestoreInternal(authHeader, request));
     }
 
-    private ListenableFuture<CompleteRestoreResponse> completeRestoreInternal(CompleteRestoreRequest request) {
+    private ListenableFuture<CompleteRestoreResponse> completeRestoreInternal(
+            AuthHeader authHeader, CompleteRestoreRequest request) {
+        if (!permittedToken.get().equals(authHeader.getBearerToken())) {
+            log.error("Attempted to complete restore with an invalid auth header", SafeArg.of("request", request));
+            throw new ServiceException(ErrorType.PERMISSION_DENIED);
```

Comment:
Should we also log the authHeader?

Reply:
We generally don't log auth headers, as we don't want to leak them.